### DISCUSSION TITLE
Worries about batching fragment shader performance

### DISCUSSION THREAD

**Discussion by o5k** [pixijs/pixijs#7537] [General]
Pixi's batching shader generator creates fragment shaders that look like this:

![](https://kagari.online/kamino_tamao.png)

Isn't this really slow? To my knowledge, these branches would be flattened to 16 texture2D calls, which doesn't sound very good (considering it's the main shader). I can't think of any way to rewrite just the shader though (can't use non-uniforms as array indices, or even just do one texture2D call outside of the ifelse chain). I'd assume it's still faster than doing one drawcall per texture, but it still sounds problematic.

Any insights on whether this is something to worry about?

**Comment by ivanpopelyshev** [comment]
Set something-something in settings to 1 and it will go away, then compare whether its good. I dont remember whether it was benchmarked or how

**Reply by o5k** [reply]
PIXI.settings.SPRITE_MAX_TEXTURES? Setting that to 1 didn't work for me, I still got a batching shader. I'd assume batching will always be faster than not batching though..

I also tried to rewrite the batching to avoid branching:
![](https://kagari.online/chinami_minami.png)
(I also made vTextureId be 10^(id). I could also rewrite this into a forloop for readability, should be OK since that'd get unrolled anyway)

This seemed slower though, which I suppose makes sense (more operations), at least in a very controlled environment where most of the time every warp will be rendering the same texture. Although, I'd believe the performance of the current branching shader would approach (or perhaps get worse than? due to the nature of branching) the rewritten one if lots of textures were used interchangably, and every warp would have to use many of them.

**Reply by ivanpopelyshev** [reply]
Oh, you just want to remove IF's? That wont help.

IF << many texture 2D, that's for sure.

There's potential memory bandwidth problem in case all those textures are 4k - that's what im concerned about. IFs on itself are working fine , @GoodBoyDigital can confirm that , he added them )

Btw, no devices give you that precision on floats in fragment shader 24bit float maximum :) `varying` on certain devices can derp it even more, texId wont be precise at all.

**Reply by ivanpopelyshev** [reply]
btw for mobile devices limit is different , i think its 4 ?

also, you have to use `settings.SPRITE_MAX_TEXTURES` before you create any renderer, because all settings params are usually copied in corresponding classes on creation (in those classes constructors)

**Reply by GoodBoyDigital** [reply]
I think branching being slow on constant things like uniforms / attributes is no longer really a thing!
I did not quite believe that my self - but it makes little difference last time i tested :D

**Reply by o5k** [reply]
Ah, I see. I just tested with just one texture max, but indeed that seems slower, makes sense as often you'll be using many textures together so you'd break pmuch any batching otherwise.

> I think branching being slow on constant things like uniforms / attributes is no longer really a thing!

I'm quite sure uniforms are free nowadays, but I can't seem to find any source on how it works with attributes.. to be fair, attributes are still probably easier for drivers to handle than truly dynamic values, so perhaps the driver can do some things beforehand? It's definitely faster than the approach I tried :p
